src/NUnitFramework/framework/Internal/Filters/NamespaceFilter.cs

```diff
@@ -0,0 +1,84 @@
+﻿// ***********************************************************************
+// Copyright (c) 2015 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using NUnit.Framework.Interfaces;
+
+namespace NUnit.Framework.Internal.Filters
+{
+    /// <summary>
+    /// ClassName filter selects tests based on the class FullName
+    /// </summary>
+#if !PORTABLE && !NETSTANDARD1_6
+    [Serializable]
+#endif
+    public class NamespaceFilter : ValueMatchFilter
+    {
+        /// <summary>
+        /// Construct a NamespaceFilter for a single namespace
+        /// </summary>
+        /// <param name="expectedValue">The namespace the filter will recognize.</param>
+        public NamespaceFilter(string expectedValue) : base(expectedValue) { }
+
+        /// <summary>
+        /// Match a test against a single value.
+        /// </summary>
+        public override bool Match(ITest test)
+        {
+            string containingNamespace = null;
+
+            if (test.TypeInfo != null)
+            {
+                containingNamespace = test.TypeInfo.Namespace;
+            }
+            else if (test is TestSuite || test is TestFixture)
+            {
+                if (test.HasChildren)
+                {
+                    containingNamespace = test.Tests[0].TypeInfo.Namespace;
+                }
+                else
+                {
+                    var outerMostClass = test.ClassName.Substring(0, test.ClassName.IndexOf("+"));
+                    containingNamespace = outerMostClass.Substring(0, outerMostClass.LastIndexOf("."));
```

Comment:
This will crash if `test.ClassName` is null, does not contain `+`, or does not contain `.`.

I don't know why `ClassName` would ever be null, but I just saw it happen. Then against I don't understand why `TypeInfo` would ever be null lol.

Could you add a test for a non-nested class in the global namespace, where the entirety of `ClassName` is something like `Foo`? An empty string should be returned for the namespace.